### ISSUE TITLE
[Training] Quality-driven data selection (5% rule)

### ISSUE THREAD

**Issue by wallscaler** [wallscaler/baobab-tts#6]
# Issue #6: [Training] Quality-Driven Data Selection (5% Rule)

## Labels
`training`, `data`, `optimization`, `priority-high`, `phase-2`

## Objective

Implement quality-driven data selection to achieve full model performance with only 5% of training data.

## Background

From [arXiv:2407.00102](https://arxiv.org/abs/2407.00102):

> "Our methods can surpass models trained on the full instruction data using only about 5% of the raw instruction tuning dataset samples."

The key insight: map data quality to a 2D space (correlation + perplexity), then select from the high-quality region.

### Quality Space Mapping

```
         ┌────────────────────────────────────┐
         │            Perplexity              │
         │     Low ◄───────────────► High     │
         │                                    │
   High  │    ★★★★★              ░░░░░░       │
         │    ★★★★★              ░░░░░░       │  ★ = Select these
Corr.    │    ★★★★★    ▒▒▒▒▒▒    ░░░░░░       │  ▒ = Maybe
         │             ▒▒▒▒▒▒                 │  ░ = Skip
   Low   │    ░░░░░░   ░░░░░░░   ░░░░░░       │
         │                                    │
         └────────────────────────────────────┘
```

### For Speech Data

Adapt the paper's approach for audio:

```python
class SpeechQualityScorer:
    """Score speech samples for quality-driven selection."""

    def __init__(self):
        # Audio-text correlation (like CLIP but for speech)
        self.alignment_model = load_alignment_model()

        # Perplexity via ASR
        self.asr_model = load_whisper()

        # Audio quality
        self.mos_predictor = load_mos_predictor()

    def score(self, audio, text) -> Tuple[float, float]:
        """Return (correlation, quality) scores."""
        # Dimension 1: Audio-text alignment
        alignment = self.alignment_model.score(audio, text)

        # Dimension 2: Composite quality
        # - ASR confidence (inverse perplexity)
        asr_result = self.asr_model.transcribe(audio)
        asr_confidence = np.mean(asr_result.token_probs)

        # - Predicted MOS
        mos = self.mos_predictor(audio)

        # - Signal quality
        snr = compute_snr(audio)
        snr_normalized = min(snr / 40, 1.0)  # Cap at 40dB

        quality = 0.4 * asr_confidence + 0.4 * (mos / 5.0) + 0.2 * snr_normalized

        return alignment, quality
```

### Selection Algorithm

```python
class QualitySelector:
    """Select top samples by quality."""

    def __init__(self, scorer: SpeechQualityScorer):
        self.scorer = scorer

    def select(
        self,
        dataset: List[Sample],
        top_percent: float = 5.0,
    ) -> List[Sample]:
        """Select top % of samples by quality."""
        # Score all samples
        scores = []
        for sample in tqdm(dataset, desc="Scoring"):
            alignment, quality = self.scorer.score(sample.audio, sample.text)
            scores.append({
                "sample": sample,
                "alignment": alignment,
                "quality": quality,
                "combined": alignment * quality,  # Product for 2D selection
            })

        # Sort by combined score
        scores.sort(key=lambda x: x["combined"], reverse=True)

        # Select top %
        n_select = int(len(scores) * top_percent / 100)
        selected = [s["sample"] for s in scores[:n_select]]

        return selected

    def select_per_language(
        self,
        dataset: List[Sample],
        top_percent: float = 5.0,
    ) -> List[Sample]:
        """Select top % within each language."""
        by_language = defaultdict(list)
        for sample in dataset:
            by_language[sample.language].append(sample)

        selected = []
        for language, samples in by_language.items():
            lang_selected = self.select(samples, top_percent)
            selected.extend(lang_selected)
            print(f"{language}: {len(lang_selected)}/{len(samples)} selected")

        return selected
```

### Curriculum Integration

Combine with D3S for progressive expansion:

```python
class QualityCurriculum:
    """Progressive quality-based curriculum."""

    def __init__(self, selector: QualitySelector):
        self.selector = selector
        self.current_percent = 5.0

    def get_data_for_epoch(self, full_dataset, epoch: int, total_epochs: int):
        """Expand data as training progresses."""
        # Start at 5%, end at 20%
        self.current_percent = 5.0 + (15.0 * epoch / total_epochs)

        return self.selector.select(full_dataset, self.current_percent)
```

## Configuration

```yaml
data:
  quality_selection:
    enabled: true
    top_percent: 5.0          # Select top 5%
    per_language: true        # Select within each language
    progressive: true         # Expand during training
    final_percent: 20.0       # End at 20%

  scoring:
    alignment_weight: 0.5
    quality_weight: 0.5
    min_snr: 20
    min_mos: 3.5
```

## Expected Results

| Data Used | Performance |
|-----------|-------------|
| 100% random | Baseline |
| 50% random | -3.2% |
| 5% quality-selected | **+1.1%** |
| 20% quality-selected | **+2.3%** |

## Files to Create

- [ ] `src/data/quality_scorer.py`
- [ ] `src/data/quality_selector.py`
- [ ] `src/data/curriculum.py`
- [ ] `tests/test_quality_selection.py`
- [ ] `configs/data/quality_selection.yaml`

## Acceptance Criteria

- [ ] Quality scoring works for speech
- [ ] 2D space mapping implemented
- [ ] Selection produces expected % of data
- [ ] Per-language selection works
- [ ] Progressive curriculum works
- [ ] Quality improvement measurable

## Testing Plan

1. **Score distribution**: Visualize 2D quality space
2. **Selection test**: Verify top 5% selected correctly
3. **Training test**: Compare 5% quality vs 5% random
4. **Ablation**: Test different % thresholds

## References

- [Quality-Driven Selection](https://arxiv.org/abs/2407.00102)
- [D3S Dynamic Sampling](https://arxiv.org/abs/2509.22115)
- [Curriculum Learning Survey](https://arxiv.org/abs/2010.13166)
